fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java

```diff
@@ -3023,6 +3017,45 @@ public CompletableFuture<Boolean> markIndexReadable(@Nonnull String indexName) {
         return markIndexReadable(getRecordMetaData().getIndex(indexName));
     }
 
+    private CompletableFuture<Boolean> checkRangeAndUpdateIndexState(Index index, byte[] indexKey, Transaction tr, boolean allowUniquePending) {
+        // An extension function to reduce markIndexReadable's complexity
+        CompletableFuture<Optional<Range>> builtFuture = firstUnbuiltRange(index);
+        CompletableFuture<Optional<RecordIndexUniquenessViolation>> uniquenessFuture = whenAllIndexUniquenessCommitChecks(index)
+                .thenCompose(vignore -> scanUniquenessViolations(index, 1).first());
+        return CompletableFuture.allOf(builtFuture, uniquenessFuture).thenApply(vignore -> {
+            Optional<Range> firstUnbuilt = context.join(builtFuture);
+            Optional<RecordIndexUniquenessViolation> uniquenessViolation = context.join(uniquenessFuture);
+
+            if (firstUnbuilt.isPresent()) {
+                throw new IndexNotBuiltException("Attempted to make unbuilt index readable" , firstUnbuilt.get(),
+                        LogMessageKeys.INDEX_NAME, index.getName(),
+                        "unbuiltRangeBegin", ByteArrayUtil2.loggable(firstUnbuilt.get().begin),
+                        "unbuiltRangeEnd", ByteArrayUtil2.loggable(firstUnbuilt.get().end),
+                        subspaceProvider.logKey(), subspaceProvider.toString(context),
+                        LogMessageKeys.SUBSPACE_KEY, index.getSubspaceKey());
+            } else if (uniquenessViolation.isPresent()) {
+                if (allowUniquePending) {
```

Comment:
I think, as @ScottDugas might have asked at one point, the normal way that we'd handle backwards compatibility here is to guard this on the store's `formatVersion` so that we don't end up writing a new index state that we can't guarantee people have the ability to read. In particular, I'd expect the behavior to be:

1. If the format version >= `UNIQUE_PENDING_FORMAT_VERSION` and `allowUniquePending`, then mark the index as readable/readable unique pending, otherwise
2. Throw a uniqueness violation

Which preserves backwards compatibility (both in behavior and in the data that is written to the index state) until the user chooses to update the format version. If we do the documentation right, that's also an opportunity for the user to see what the new feature means for them, and let them know if they are currently relying on unique indexes always being unique (or not scannable) that they need to adjust their assumptions.

One other advantage of using the format version is that we could (if we wanted to) remove "allow unique pending" entirely from the indexer state, and just always allow unique pending in index builds. Or maybe we want it as part of that configuration in case someone _wants_ index builds with uniqueness violations to fail, but with the format version, the user's decision to allow or not allow that could be de-coupled with their deploy a bit more.

Reply:
Would making it the default format version (=exceeding HEADER_USER_FIELDS_FORMAT_VERSION) be acceptable?